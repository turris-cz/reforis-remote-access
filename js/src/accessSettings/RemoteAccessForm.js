/*
 * Copyright (C) 2019-2021 CZ.NIC z.s.p.o. (https://www.nic.cz/)
 *
 * This is free software, licensed under the GNU General Public License v3.
 * See /LICENSE for more information.
 */

import React, { useEffect } from "react";
import PropTypes from "prop-types";

import {
    CheckBox,
    NumberInput,
    Button,
    useForm,
    API_STATE,
    useAlert,
    useAPIPut,
    undefinedIfEmpty,
} from "foris";

import API_URLs from "API";

RemoteAccessForm.propTypes = {
    settings: PropTypes.object.isRequired,
    onSuccess: PropTypes.func.isRequired,
};

export default function RemoteAccessForm({ settings, onSuccess }) {
    const [setAlert] = useAlert();

    const [putSettingsResponse, putSettings] = useAPIPut(API_URLs.settings);
    useEffect(() => {
        if (putSettingsResponse.state === API_STATE.ERROR) {
            setAlert(_("Cannot save connection parameters."));
        } else if (putSettingsResponse.state === API_STATE.SUCCESS) {
            onSuccess();
        }
    }, [putSettingsResponse, setAlert, onSuccess]);

    const [formState, formChangeHandler, reloadForm] = useForm(validator);
    const formData = formState.data;
    const formErrors = formState.errors || {};
    useEffect(() => {
        reloadForm(settings);
    }, [reloadForm, settings]);

    if (!formData) {
        return null;
    }

    function handleSubmit(event) {
        event.preventDefault();
        if (formData.enabled) {
            putSettings({ data: formData });
        } else {
            putSettings({ data: { enabled: false } });
        }
    }

    const saveButtonDisabled =
        !!undefinedIfEmpty(formErrors) ||
        putSettingsResponse.state === API_STATE.SENDING;

    return (
        <>
            <h2>{_("Remote Access")}</h2>
            <form onSubmit={handleSubmit}>
                <Switch
                    label={_("Enable Remote Access")}
                    checked={formData.enabled}
                    onChange={formChangeHandler((value) => ({
                        enabled: { $set: value },
                    }))}
                />
                {formData.enabled && (
                    <>
                        <CheckBox
                            label={_("Accessible via WAN")}
                            helpText={_(`Devices in the WAN network will be \
able to connect to the configuration interface. Otherwise only devices on LAN \
will be able to access it.`)}
                            checked={formData.wan_access}
                            onChange={formChangeHandler((value) => ({
                                wan_access: { $set: value },
                            }))}
                        />
                        <NumberInput
                            label={_("Port")}
                            helpText={_(`A port which will be opened for the \
remote configuration of this device.`)}
                            value={formData.port}
                            error={formErrors.port}
                            onChange={formChangeHandler((value) => ({
                                port: { $set: value },
                            }))}
                        />
                    </>
                )}
                <Button
                    type="submit"
                    forisFormSize
                    disabled={saveButtonDisabled}
                    loading={putSettingsResponse.state === API_STATE.SENDING}
                >
                    {_("Save")}
                </Button>
            </form>
        </>
    );
}

function validator(formData) {
    if (formData.port < 1024 || formData.port > 65536) {
        return { port: _("Should be a number in range of 1024-65536") };
    }
    return undefined;
}
