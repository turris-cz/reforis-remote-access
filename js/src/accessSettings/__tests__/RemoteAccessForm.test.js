/*
 * Copyright (C) 2019 CZ.NIC z.s.p.o. (http://www.nic.cz/)
 *
 * This is free software, licensed under the GNU General Public License v3.
 * See /LICENSE for more information.
 */

import React from "react";

import mockAxios from "jest-mock-axios";
import {
    render,
    fireEvent,
    getByLabelText,
    getByText,
    queryByText,
    wait,
    getByTestId,
} from "foris/testUtils/customTestRender";
import { mockSetAlert } from "foris/testUtils/alertContextMock";
import { mockJSONError } from "foris/testUtils/network";

import RemoteAccessForm from "../RemoteAccessForm";

describe("<RemoteAccessForm />", () => {
    let container;
    const onSuccess = jest.fn();
    const settings = { enabled: true, wan_enabled: true, port: 1808 };
    const settingsDisabled = { enabled: false, wan_enabled: false, port: 1194 };

    function renderForm(settings) {
        return render(
            <RemoteAccessForm settings={settings} onSuccess={onSuccess} />
        );
    }

    function getSaveButton() {
        return getByTestId(container, "save-settings-button");
    }

    function saveForm() {
        fireEvent.click(getSaveButton());
    }

    function enableRemoteAccess() {
        fireEvent.click(getByLabelText(container, "Enable Remote Access"));
    }

    function setPort(port) {
        fireEvent.change(getByLabelText(container, "Port"), {
            target: { value: port },
        });
    }

    it("should render only checkbox when access is disabled", () => {
        ({ container } = renderForm(settingsDisabled));
        expect(container).toMatchSnapshot();
    });

    it("should render full form when access is enabled", () => {
        ({ container } = renderForm(settings));
        expect(container).toMatchSnapshot();
    });

    it("should expose the rest of the form when enabled", () => {
        ({ container } = renderForm(settingsDisabled));
        enableRemoteAccess();
        expect(container).toMatchSnapshot();
    });

    it("should validate port", () => {
        ({ container } = renderForm(settings));

        // Invalid ports
        for (const port of [-1, 9999999999999999, 80]) {
            setPort(port);
            expect(getSaveButton().disabled).toBeTruthy();
            expect(
                getByText(
                    container,
                    "Should be a number in range of 1024-65536"
                )
            ).toBeTruthy();
        }

        // Valid port
        setPort(9090);
        expect(getSaveButton().disabled).toBeFalsy();
        expect(
            queryByText(container, "Should be a number in range of 1024-65536")
        ).toBeNull();
    });

    it("should handle success", async () => {
        ({ container } = renderForm(settingsDisabled));

        // Fill out and submit the form
        enableRemoteAccess();
        const newPort = 9090;
        setPort(newPort);
        saveForm();

        expect(mockAxios.put).toBeCalledWith(
            "/reforis/remote-access/api/settings",
            { ...settingsDisabled, enabled: true, port: newPort },
            expect.anything()
        );

        mockAxios.mockResponse();
        await wait(() => expect(onSuccess).toBeCalled());
    });

    it("should turn off remote access", async () => {
        ({ container } = renderForm(settings));
        // Fill out and submit the form
        enableRemoteAccess();
        saveForm();
        expect(mockAxios.put).toBeCalledWith(
            "/reforis/remote-access/api/settings",
            { enabled: false },
            expect.anything()
        );
    });

    it("should set alert on API error", async () => {
        ({ container } = renderForm(settings));
        saveForm();
        mockJSONError();
        await wait(() =>
            expect(mockSetAlert).toBeCalledWith(
                "Cannot save connection parameters."
            )
        );
    });

    it("should display a spinner and disable save button while sending request", async () => {
        ({ container } = renderForm(settings));
        saveForm();
        expect(getSaveButton().disabled).toBeTruthy();
        expect(container).toMatchSnapshot();
    });
});
