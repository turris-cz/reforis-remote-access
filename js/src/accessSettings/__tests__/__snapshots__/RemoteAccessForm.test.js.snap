// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<RemoteAccessForm /> should display a spinner and disable save button while sending request 1`] = `
<div>
  <div
    class="card p-4 col-sm-12 col-lg-12 p-0 mb-4"
  >
    <h2>
      Remote Access
    </h2>
    <form>
      <div
        class="form-check form-switch mb-3"
      >
        <input
          checked=""
          class="form-check-input"
          id="1"
          role="switch"
          type="checkbox"
        />
        <label
          class="form-check-label"
          for="1"
        >
          Enable Remote Access
        </label>
      </div>
      <div
        class="form-check form-switch mb-3"
      >
        <input
          class="form-check-input"
          id="2"
          role="switch"
          type="checkbox"
        />
        <label
          class="form-check-label"
          for="2"
        >
          Accessible via WAN
        </label>
        <div
          class="form-text"
        >
          <small>
            Devices in the WAN network will be able to connect to the configuration interface. Otherwise, only devices on LAN will be able to access it.
          </small>
        </div>
      </div>
      <div
        class="mb-3"
      >
        <label
          class="form-label"
          for="3"
        >
          Port
        </label>
        <div
          class="input-group"
        >
          <input
            class="form-control"
            id="3"
            type="number"
            value="1808"
          />
          <button
            aria-label="Increase value. Hint: Hold to increase faster."
            class="btn btn-outline-secondary"
            title="Increase value. Hint: Hold to increase faster."
            type="button"
          >
            <i
              class="fa"
            />
          </button>
          <button
            aria-label="Decrease value. Hint: Hold to decrease faster."
            class="btn btn-outline-secondary"
            title="Decrease value. Hint: Hold to decrease faster."
            type="button"
          >
            <i
              class="fa"
            />
          </button>
        </div>
        <div
          class="form-text"
        >
          <small>
            A port which will be opened for the remote configuration of this device.
          </small>
        </div>
      </div>
      <div
        class="text-end"
      >
        <button
          class="btn btn-primary col-12 col-md-3 col-lg-2 d-inline-flex justify-content-center align-items-center"
          data-testid="save-settings-button"
          disabled=""
          type="submit"
        >
          <span
            aria-hidden="true"
            class="spinner-border spinner-border-sm me-1"
            role="status"
          />
          <span>
            Save
          </span>
        </button>
      </div>
    </form>
  </div>
</div>
`;

exports[`<RemoteAccessForm /> should expose the rest of the form when enabled 1`] = `
<div>
  <div
    class="card p-4 col-sm-12 col-lg-12 p-0 mb-4"
  >
    <h2>
      Remote Access
    </h2>
    <form>
      <div
        class="form-check form-switch mb-3"
      >
        <input
          class="form-check-input"
          id="1"
          role="switch"
          type="checkbox"
        />
        <label
          class="form-check-label"
          for="1"
        >
          Enable Remote Access
        </label>
      </div>
      <div
        class="form-check form-switch mb-3"
      >
        <input
          class="form-check-input"
          id="2"
          role="switch"
          type="checkbox"
        />
        <label
          class="form-check-label"
          for="2"
        >
          Accessible via WAN
        </label>
        <div
          class="form-text"
        >
          <small>
            Devices in the WAN network will be able to connect to the configuration interface. Otherwise, only devices on LAN will be able to access it.
          </small>
        </div>
      </div>
      <div
        class="mb-3"
      >
        <label
          class="form-label"
          for="3"
        >
          Port
        </label>
        <div
          class="input-group"
        >
          <input
            class="form-control"
            id="3"
            type="number"
            value="1194"
          />
          <button
            aria-label="Increase value. Hint: Hold to increase faster."
            class="btn btn-outline-secondary"
            title="Increase value. Hint: Hold to increase faster."
            type="button"
          >
            <i
              class="fa"
            />
          </button>
          <button
            aria-label="Decrease value. Hint: Hold to decrease faster."
            class="btn btn-outline-secondary"
            title="Decrease value. Hint: Hold to decrease faster."
            type="button"
          >
            <i
              class="fa"
            />
          </button>
        </div>
        <div
          class="form-text"
        >
          <small>
            A port which will be opened for the remote configuration of this device.
          </small>
        </div>
      </div>
      <div
        class="text-end"
      >
        <button
          class="btn btn-primary col-12 col-md-3 col-lg-2 d-inline-flex justify-content-center align-items-center"
          data-testid="save-settings-button"
          type="submit"
        >
          <span>
            Save
          </span>
        </button>
      </div>
    </form>
  </div>
</div>
`;

exports[`<RemoteAccessForm /> should render full form when access is enabled 1`] = `
<div>
  <div
    class="card p-4 col-sm-12 col-lg-12 p-0 mb-4"
  >
    <h2>
      Remote Access
    </h2>
    <form>
      <div
        class="form-check form-switch mb-3"
      >
        <input
          checked=""
          class="form-check-input"
          id="1"
          role="switch"
          type="checkbox"
        />
        <label
          class="form-check-label"
          for="1"
        >
          Enable Remote Access
        </label>
      </div>
      <div
        class="form-check form-switch mb-3"
      >
        <input
          class="form-check-input"
          id="2"
          role="switch"
          type="checkbox"
        />
        <label
          class="form-check-label"
          for="2"
        >
          Accessible via WAN
        </label>
        <div
          class="form-text"
        >
          <small>
            Devices in the WAN network will be able to connect to the configuration interface. Otherwise, only devices on LAN will be able to access it.
          </small>
        </div>
      </div>
      <div
        class="mb-3"
      >
        <label
          class="form-label"
          for="3"
        >
          Port
        </label>
        <div
          class="input-group"
        >
          <input
            class="form-control"
            id="3"
            type="number"
            value="1808"
          />
          <button
            aria-label="Increase value. Hint: Hold to increase faster."
            class="btn btn-outline-secondary"
            title="Increase value. Hint: Hold to increase faster."
            type="button"
          >
            <i
              class="fa"
            />
          </button>
          <button
            aria-label="Decrease value. Hint: Hold to decrease faster."
            class="btn btn-outline-secondary"
            title="Decrease value. Hint: Hold to decrease faster."
            type="button"
          >
            <i
              class="fa"
            />
          </button>
        </div>
        <div
          class="form-text"
        >
          <small>
            A port which will be opened for the remote configuration of this device.
          </small>
        </div>
      </div>
      <div
        class="text-end"
      >
        <button
          class="btn btn-primary col-12 col-md-3 col-lg-2 d-inline-flex justify-content-center align-items-center"
          data-testid="save-settings-button"
          type="submit"
        >
          <span>
            Save
          </span>
        </button>
      </div>
    </form>
  </div>
</div>
`;

exports[`<RemoteAccessForm /> should render only checkbox when access is disabled 1`] = `
<div>
  <div
    class="card p-4 col-sm-12 col-lg-12 p-0 mb-4"
  >
    <h2>
      Remote Access
    </h2>
    <form>
      <div
        class="form-check form-switch mb-3"
      >
        <input
          class="form-check-input"
          id="1"
          role="switch"
          type="checkbox"
        />
        <label
          class="form-check-label"
          for="1"
        >
          Enable Remote Access
        </label>
      </div>
      <div
        class="text-end"
      >
        <button
          class="btn btn-primary col-12 col-md-3 col-lg-2 d-inline-flex justify-content-center align-items-center"
          data-testid="save-settings-button"
          type="submit"
        >
          <span>
            Save
          </span>
        </button>
      </div>
    </form>
  </div>
</div>
`;
